src/mqtt/protocol/mqtt/mqtt_client.c

```diff
@@ -0,0 +1,1110 @@
+
+	default:
+		// unexpected packet type, server misbehaviour
+		nni_mtx_unlock(&s->mtx);
+		nni_pipe_close(p->pipe);
+		return;
+	}
+
+	if (work_is_error(work)) {
+		// protocol error, just close the connection
+		nni_mtx_unlock(&s->mtx);
+		nni_aio_finish_error(work->user_aio, NNG_EPROTO);
+		nni_pipe_close(p->pipe);
+		return;
+	} else if (work_is_final(work)) {
+		// good news, protocol state machine run to the end
+		nni_aio *aio = work->user_aio;
+		mqtt_sock_free_work(s, work);
+		nni_mtx_unlock(&s->mtx);
+		nni_aio_finish(aio, 0, 0);
+		return;
+	}
+
+	nni_mtx_unlock(&s->mtx);
+
+	return;
+}
+
+// Note: This routine should be called with the sock lock held.
+static void
+mqtt_recv_start(mqtt_sock_t *s, nni_aio *aio)
+{
+	mqtt_pipe_t *p = s->mqtt_pipe;
+	work_t *     work;
+
+	work = mqtt_sock_get_work(s);
+
+	if (NULL == work) {
+		nni_mtx_unlock(&s->mtx);
+		nni_aio_finish_error(aio, NNG_ENOMEM);
+		return;
+	}
+
+	work->user_aio = aio;
+	work_set_recv(work, NNG_MQTT_PUBLISH);
+	nni_list_append(&s->recv_queue, work); // enqueue to recv
+
+	// no open pipe
+	if (NULL == p || nni_atomic_get_bool(&p->closed)) {
+		return;
+	}
+
+	mqtt_run_recv_queue(s);
+}
+
+// Note: This routine should be called with the sock lock held.
+static void
+mqtt_run_recv_queue(mqtt_sock_t *s)
+{
+	work_t *     work = nni_list_first(&s->recv_queue);
+	mqtt_pipe_t *p    = s->mqtt_pipe;
+	nni_msg *    msg;
+
+	while (NULL != work) {
+		if (0 != nni_lmq_get(&p->recv_messages, &msg)) {
+			break;
+		}
+		nni_list_remove(&s->recv_queue, work);
+		// nni_pipe_recv(p->pipe, &work->recv_aio);
+		nni_aio_set_msg(work->user_aio, msg);
+		nni_aio_finish(work->user_aio, 0,
+		    nni_msg_header_len(msg) + nni_msg_len(msg));
+		mqtt_sock_free_work(s, work);
+		work = nni_list_first(&s->recv_queue);
+	}
+
+	return;
+}
+
+// Note: This routine should be called with the sock lock held.
+static void
+mqtt_send_start(mqtt_sock_t *s)
+{
+	work_t *     work;
+	mqtt_pipe_t *p = s->mqtt_pipe;
+	uint16_t     packet_type;
+
+	// no open pipe
+	if (NULL == p) {
+		return;
+	}
+
+	if (NULL != (work = nni_list_first(&s->send_queue))) {
+		packet_type = nni_mqtt_msg_get_packet_type(work->msg);
+
+		// we are not retransmitting the work
+		if (0 == work->ntrial) {
+			// only allow to send CONNECT, PINGREQ, PUBLISH,
+			// SUBSCRIBE and UNSUBSCRIBE
+			switch (packet_type) {
+			case NNG_MQTT_CONNECT:
+				// NOTE: the transport dialer handle CONNECT
+				// fall through
+			case NNG_MQTT_PINGREQ:
+				break;
+
+			case NNG_MQTT_PUBLISH:
+				work->qos =
+				    nni_mqtt_msg_get_publish_qos(work->msg);
+				if (0 == work->qos) {
+					break; // QoS 0 need no packet id
+				}
+				// fall through
+			case NNG_MQTT_SUBSCRIBE:
+			case NNG_MQTT_UNSUBSCRIBE:
+				work->packet_id =
+				    mqtt_pipe_get_next_packet_id(p);
+				nni_mqtt_msg_set_packet_id(
+				    work->msg, work->packet_id);
+				NNI_ASSERT(nni_id_get(&p->send_unack,
+				               work->packet_id) == NULL);
+				if (0 !=
```

Comment:
Style:  I *hate* this style.  Please put the constant *after* the condition.  And if this is a boolean condition, please just use boolean tests, rather than comparing against integers.

Reply:
removed